components/sup/src/manager/service/mod.rs

```diff
@@ -407,29 +412,46 @@ impl Service {
         self.supervisor.state == ProcessState::Down
     }
 
-    /// Compares the current state of the service to the current state of the census ring and
-    /// re-renders all templatable content to disk.
+    /// Compares the current state of the service to the current state of the census ring and the
+    /// user-config, and re-renders all templatable content to disk.
     ///
-    /// Returns true if any modifications were made.
+    /// Returns `true` if any modifications were made.
     fn update_templates(&mut self, census_ring: &CensusRing) -> bool {
         let census_group = census_ring.census_group_for(&self.service_group).expect(
             "Service update failed; unable to find own service group",
         );
-        let cfg_updated = self.cfg.update(census_group);
-        if cfg_updated || census_ring.changed() {
-            let (reload, reconfigure) = {
-                let ctx = self.render_context(census_ring);
-                let reload = self.compile_hooks(&ctx);
-                let reconfigure = self.compile_configuration(&ctx);
-                (reload, reconfigure)
-            };
-            self.needs_reload = reload;
-            self.needs_reconfiguration = reconfigure;
+        let cfg_updated_from_rumors = self.cfg.update(census_group);
+
+        if self.user_config_updated {
+            if let Err(e) = self.cfg.reload_user() {
+                outputln!(preamble self.service_group, "Loading user-config failed: {}", e);
+            }
         }
-        cfg_updated
+
+        let (cfg_changed, reload, reconfigure) = {
+            let ctx = self.render_context(census_ring);
+
+            if cfg_updated_from_rumors || census_ring.changed() {
+                let config_changed = cfg_updated_from_rumors || self.user_config_updated;
+                let needs_reload = self.compile_hooks(&ctx) || self.user_config_updated;
+                let needs_reconfigure = self.compile_configuration(&ctx);
+
+                (config_changed, needs_reload, needs_reconfigure)
+            } else if self.user_config_updated {
+                (true, true, self.compile_configuration(&ctx))
+            } else {
+                (false, self.needs_reload, self.needs_reconfiguration)
```

Comment:
I believe that `needs_reload` and `needs_reconfiguration` are only set inside this function, so I think we can collapse this branch of the `if/else`, too.

Reply:
I'm not sure I understand, but I'll bite - `needs_reload` and `needs_reconfiguration` are set to false also in `reload` and `reconfigure` functions, which are also gated on `needs_reload` and `needs_reconfiguration` being true. The `(false, self.needs_reload, self.needs_reconfiguration)` tuple here is to try to avoid changing value any of the members, so we don't accidentally set them from `true` to `false` because some config we don't care about has changed, but the reloading/reconfiguring didn't yet happen.

I haven't investigated if this can be the case, though.

Reply:
See reply above... I don't think additional if/else comparisons should be needed if new user config is accounted for when we generate our template rendering context.